### PR TITLE
Add `Concurrent.cpu_requests` that is cgroups aware.

### PR DESCRIPTION
**Pull request by heka1024**
In K8s environments, cpu requests are often more useful than limits, but concurrent-ruby currently lacks a related method, so I've added it. Below are links to articles about the usefulness of cpu requests.

- https://medium.com/directeam/kubernetes-resources-under-the-hood-part-3-6ee7d6015965
- https://x.com/thockin/status/1134193838841401345

### CODE REVIEW

**Comment by eregon** [comment]
Could you summarize in one/a few sentences why one should use cpu requests instead of cpu limits?

**Comment by heka1024** [comment]
@eregon CPU limits can lead to throttling and reduced performance under high load. So, as [Tim Hockin](https://x.com/thockin) suggested, I've set only requests in my application at company.

File: lib/concurrent-ruby/concurrent/utility/processor_counter.rb [hunk 1]
```diff
@@ -187,4 +212,12 @@ def self.available_processor_count
   def self.cpu_quota
     processor_counter.cpu_quota
   end
+
+  # The CPU shares requested by the process. For performance reasons the calculated
+  # value will be memoized on the first call.
+  #
+  # @return [Float, nil] CPU shares requested by the process, or nil if not set
+  def self.cpu_requests
```
Comment (eregon) [inline lib/concurrent-ruby/concurrent/utility/processor_counter.rb:220]: I wonder what's the best name here.
`cpu_shares` might be better since it reads from `/sys/fs/cgroup/cpu.weight` or `/sys/fs/cgroup/cpu/cpu.shares`.
And we already have `cpu_quota` which reads from `cpu.cfs_quota_us`/`cpu.max`.

`requests` sounds like a Kubernetes-specific term, notably Docker calls it shares in https://docs.docker.com/config/containers/resource_constraints/#cpu

So could you rename to `cpu_shares`?

File: lib/concurrent-ruby/concurrent/utility/processor_counter.rb [hunk 2]
```diff
@@ -41,6 +42,10 @@ def cpu_quota
         @cpu_quota.value
       end
 
+      def cpu_requests
+        @cpu_requests.value / 1024
```
Comment (eregon) [inline lib/concurrent-ruby/concurrent/utility/processor_counter.rb:46]: This will NoMethodError if there is neither cgroup 1 or 2.
I think it makes sense to move the `/ 1024` in `compute_cpu_requests`

**Comment by eregon** [comment]
@byroot @casperisfine Could you review this too?

File: lib/concurrent-ruby/concurrent/utility/processor_counter.rb (hunk 1 above)
Comment (heka1024) [reply lib/concurrent-ruby/concurrent/utility/processor_counter.rb]: @eregon Thank you for suggestion. I've made changes based on comment.

**Review by eregon — APPROVED** [review]

File: lib/concurrent-ruby/concurrent/utility/processor_counter.rb [hunk 3]
```diff
@@ -113,6 +118,26 @@ def compute_cpu_quota
           end
         end
       end
+
+      def compute_cpu_shares
+        if RbConfig::CONFIG["target_os"].include?("linux")
+          if cgroups_v2?
+            # Ref: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2254-cgroup-v2#phase-1-convert-from-cgroups-v1-settings-to-v2
+            weight = File.read('/sys/fs/cgroup/cpu.weight').to_f
```
Comment (byroot) [inline lib/concurrent-ruby/concurrent/utility/processor_counter.rb:126]: It's probably fine, but in `compute_cpu_quota` we only ever read files we tested for existence prior.

Here we check `/sys/fs/cgroup/cpu.max` and assume it means `/sys/fs/cgroup/cpu.weight` exists? Maybe that's a correct assertion, but I personally don't know if it's true.

I think ultimately it would be simpler to not extract `cgroups_v1? / cgroups_v2?` and check if `/sys/fs/cgroup/cpu.weight` exists here.

File: lib/concurrent-ruby/concurrent/utility/processor_counter.rb [hunk 4]
```diff
@@ -113,6 +118,26 @@ def compute_cpu_quota
...
+            ((((weight - 1) * 262142) / 9999) + 2) / 1024
+          elsif cgroups_v1?
+            File.read('/sys/fs/cgroup/cpu/cpu.shares').to_f / 1024
```
Comment (byroot) [inline lib/concurrent-ruby/concurrent/utility/processor_counter.rb:129]: Same issue here.

**Review by byroot — APPROVED** [review]
I'd undo the "extract method" for extra safety, but other than that LGTM.

File: lib/concurrent-ruby/concurrent/utility/processor_counter.rb (hunk 3 above)
Comment (eregon) [reply lib/concurrent-ruby/concurrent/utility/processor_counter.rb:126]: Agreed, this seems safer

**Comment by heka1024** [comment]
@byroot @heka1024 I'd undo the "extract method" and replace it with file existence check. Could you review this once again?

**Review by byroot — APPROVED** [review]

**Review by eregon — APPROVED** [review]
